packages/core/_test/stm/STM/taps.test.ts

```diff
@@ -46,42 +46,51 @@ describe.concurrent("STM", () => {
-      // val tx =
-      //   for {
-      //     tapSuccess    <- TPromise.make[Nothing, Int]
-      //     tapError      <- TPromise.make[Nothing, String]
-      //     succeededSTM   = ZSTM.succeed(42): STM[String, Int]
-      //     result        <- succeededSTM.tapBoth(e => tapError.succeed(e), a => tapSuccess.succeed(a))
-      //     tappedSuccess <- tapSuccess.await
-      //   } yield (result, tappedSuccess)
-      // assertM(tx.commit)(equalTo((42, 42)))
+      const tx = Do(($) => {
+        const tapSuccess = $(TDeferred.make<never, number>())
+        const tapError = $(TDeferred.make<never, string>())
+        const succeededSTM = STM.succeed(42)
+        const result = $(succeededSTM.tapBoth(e => tapError.succeed(e), a => tapSuccess.succeed(a)))
+        const tappedSuccess = $(tapSuccess.await)
+
+        return result === 42 && tappedSuccess === 42
+      }).commit()
+
+      const result = await tx.unsafeRunPromise()
+
+      assert.isTrue(result)
     })
 
-    // TODO: implement after TPromise
     it("tapBoth applies the function to error and successful values while keeping the effect itself on error", async () => {
-      // val tx =
-      //   for {
-      //     tapSuccess  <- TPromise.make[Nothing, Int]
-      //     tapError    <- TPromise.make[Nothing, String]
-      //     succeededSTM = ZSTM.fail("error"): STM[String, Int]
-      //     result      <- succeededSTM.tapBoth(e => tapError.succeed(e), a => tapSuccess.succeed(a)).either
-      //     tappedError <- tapError.await
-      //   } yield (result, tappedError)
-      // assertM(tx.commit)(equalTo((Left("error"), "error")))
-    })
+      const tx = Do(($) => {
+        const tapSuccess = $(TDeferred.make<never, number>())
+        const tapError = $(TDeferred.make<never, string>())
+        const succeededSTM: STM<never, string, number> = STM.fail("error")
+        const result = $(succeededSTM.tapBoth(e => tapError.succeed(e), a => tapSuccess.succeed(a)).either)
+        const tappedError = $(tapError.await)
+
+        return result == Either.left("error") && tappedError === "error"
+      }).commit()
 
-    // TODO: implement after TPromise
-    it("tapError should apply the transactional function to the error result while keeping the effect itself", async () => {
-      // val tx =
-      //   for {
-      //     errorRef    <- TPromise.make[Nothing, String]
-      //     failedStm    = ZSTM.fail("error") *> ZSTM.succeed(0)
-      //     result      <- failedStm.tapError(e => errorRef.succeed(e)).either
-      //     tappedError <- errorRef.await
-      //   } yield (result, tappedError)
-      // assertM(tx.commit)(equalTo((Left("error"), "error")))
+      const result = await tx.unsafeRunPromise()
+
+      assert.isTrue(result)
     })
+
+    // it("tapError should apply the transactional function to the error result while keeping the effect itself", async () => {
```

Comment:
@mikearnaldi This test is failing expectedly (https://github.com/zio/zio/blob/a822917ad730910fd69f38c341ab134995cfae95/core-tests/shared/src/test/scala/zio/stm/ZSTMSpec.scala#L1197), but I can't seem to find a way to indicate this in a way that it actually doesn't crash and burn.